test/e2e/mcd_test.go

```diff
@@ -533,3 +534,46 @@ func TestCustomPool(t *testing.T) {
 	err = waitForPoolComplete(t, cs, "infra", renderedConfig)
 	require.Nil(t, err)
 }
+
+func TestIgn3Cfg(t *testing.T) {
+	cs := framework.NewClientSet("")
+
+	// create a dummy MC with an sshKey for user Core
+	mcName := fmt.Sprintf("ign3cfg-worker-%s", uuid.NewUUID())
+	mcadd := &mcfgv1.MachineConfig{}
+	mcadd.ObjectMeta = metav1.ObjectMeta{
+		Name:   mcName,
+		Labels: mcLabelForWorkers(),
+	}
+	// create a new MC that adds a valid user & ssh key
+	Ign3Config := ign3types.Config{}
+	tempUser := ign3types.PasswdUser{Name: "core", SSHAuthorizedKeys: []ign3types.SSHAuthorizedKey{"1234_test_ign3"}}
+	Ign3Config.Passwd.Users = append(Ign3Config.Passwd.Users, tempUser)
+	Ign3Config.Ignition.Version = "3.0.0"
+	rawIgnConfig := helpers.MarshalOrDie(Ign3Config)
+	mcadd.Spec.Config.Raw = rawIgnConfig
+
+	_, err := cs.MachineConfigs().Create(context.TODO(), mcadd, metav1.CreateOptions{})
```

Comment:
This is failing because the CRD manifest for MachineConfig is still v2 (manifests/machineconfig.crd.yaml) and it errors out because ignition.config.replace.source is a pointer to string in v3 which can be nullable. I'm not sure what we can do here to allow both as we're now allowing v2 and v3 and they have differences 🤔 @yuqi-zhang do you have any idea if we can allow that? or just make the string nullable as a pointer?

Reply:
I think this should be fixable by adding a `nullable: true` after https://github.com/openshift/machine-config-operator/blob/master/manifests/machineconfig.crd.yaml#L87 , in parallel with `source`. Example usage: https://github.com/openshift/machine-config-operator/blob/master/manifests/machineconfig.crd.yaml#L316

If that does not work we can just drop the field altogether.

To be precise crd validation happens the same way when you oc create -f I think. My best guess is that Create() gets a full v3 object validated whereas oc create -f just assumes its a spec 2 object which is processed later.

Reply:
We may see other locations with this error since a lot of spec 3 uses pointers, so you may have to fix as we go. We can compare between https://github.com/openshift/machine-config-operator/blob/c8e9ccf5569ae05b2bdaf3a44c5e371ebb5e6ade/vendor/github.com/coreos/ignition/config/v2_3/types/schema.go and https://github.com/openshift/machine-config-operator/blob/c8e9ccf5569ae05b2bdaf3a44c5e371ebb5e6ade/vendor/github.com/coreos/ignition/v2/config/v3_0/types/schema.go

Reply:
Is this really the cause of the problem here? The diff for the CRD in the fcos branch is empty (`git diff fcos manifests/machineconfig.crd.yaml`)

Reply:
that is to say the CRD validation works as-is on Spec3 in OKD

Reply:
I think we hit a very similar same issue in installer:fcos recently: The marshaling for empty pointer structs didn't behave as desired and they were included in the marshalled data as weird nil fields. Our solution for OKD (which has since also been adopted in the fedora-config-transpiler as well) was to switch out the standard json.Marshal with a drop in replacement from `github.com/clarketm/json` which supports supports zero values of structs
with omitempty markers when marshaling them (it'll drop those fields).
This may be a better solution than regressing on openAPI spec validation.

for reference: https://github.com/coreos/fcct/pull/90

Reply:
@lorbus yours ARE different you don't have jerry's latest chagnes:
https://github.com/openshift/machine-config-operator/commits/fcos/manifests/machineconfig.crd.yaml

https://github.com/openshift/machine-config-operator/commits/master/manifests/machineconfig.crd.yaml

The last commit you have that matches the MCO is from December =)

Reply:
From my count that's 5 commits your branch is missing..

Reply:
We don't currently preserve commit history on the fcos branch to include changes with merges instead of rebases. Did you run the diff cmd? (I just double checked and the diff is empty for me)

Reply:
didnt diff, i assumed you pulled commit history. would be good to keep it since to preserve authors..

Reply:
it would be good indeed, but it was a compromise we had to make to catch up on the ever growing gap between fcos and master (fixing up every commit's merge errors is much more work than just checking out the state you want on top of current commit. And with the fcos branch being temporary all of that work would be thrown away later anyway. we took that route route.) Hopefully we won't have to maintain that duality much longer now :)

Reply:
I don't think this is resolved -- I believe the problem could be fixed by using a marshaler that excludes the empty pointer struct fields from the data, then a regression on the OpenAPI validation wouldn't be needed (as I mentioned before we do that on OKD already withe exactly the same validation theme used for in OCP right now, compare: https://github.com/openshift/machine-config-operator/blob/fcos/test/helpers/helpers.go#L4)

Reply:
The above will become necessary when moving the CL config yamls to FCCT format in order to produce valid marshaled data. As this is out of scope for this PR, I'll mark as resolved.